### DISCUSSION TITLE
Subscriptable API for `limit` and `offset` in queries

### DISCUSSION THREAD

**Discussion by Alc-Alc** [sqlalchemy/sqlalchemy#11813] [Ideas]
SQLAlchemy provides a rich variety of overloads to write "pythonic sql queries", the following are few examples of this.

### Comparison
`select(SomeModel).where(SomeModel.id < 2)`

### Negation
`select(SomeModel).where(~SomeModel.name.startswith('something'))`

### Conditionals (via bitwise operators)
`select(SomeModel).where((SomeModel.id < 2) & (~SomeModel.name.startswith('something')))`

---
With this context, I would like to know if an overload for `__getitem__` will be entertained for `limit` and `offset`. Note that these are not meant to replace the current APIs but rather coexist with their method  counterparts. The examples shown here are for `select` but I would expect these to also apply anywhere `.limit` and `.offset` can be used.

### Apply only `limit`

Current API - `select(SomeModel).limit(5)`

Proposed API - `select(SomeModel)[:5]`

IOW, `slice.start` is `None` and `slice.stop` is `5`.

####  Proposed API - `select(SomeModel)[2:]`

IOW, `slice.start` is `2` and `slice.stop` is `None`.

#### Proposed API 1  - `select(SomeModel)[2:5]`

IOW, `slice.start` is `2` and `slice.stop` is `5`.

"Proposed API 1" is more direct (than the one that follows) in getting the limits and offsets across but it might be considered "more magic" when compared to regular Python list / tuple behavior. More on this below.

####  Proposed API 2 - `select(SomeModel)[2:7]`

IOW, `slice.start` is `2` and `slice.stop` is `7`.

Note that `slice.stop` is `7` (not `5`). The reason being similar to how slicing a list / tuple works. It can be considered intuitive but it requires some calculations to get the offsets correct.

PS: I will be also be content with a comment explaining why this is not favorable 🙂
PPS: Pagination is indeed hard 🙂 but it would help (IMO) to have an easier API.

**Comment by zzzeek** [comment]
we have this for orm.Query().    if you use `Query` instead of `select()`, it will do this.   however, it also runs the SQL at that point and returns a list, since we felt that actually asking for a slice should be returning a list and not an iterable.

I'm -1 on sticking this straight onto a `select()` since it feels wrong for a slice operator to return something that isn't a list.   

instead, use the [slice() method directly](https://docs.sqlalchemy.org/en/20/core/selectable.html#sqlalchemy.sql.expression.Select.slice).  this makes it clear that this is not a list-returning method and you can apply the same slice-like semantics to your values.

**Reply by Alc-Alc** [reply]
thank you for the prompt response, I was not aware of `slice` in SQLAlchemy. This issue can be closed.

**Reply by CaselIT** [reply]
There's a typo in the slice example.

I'm also -1 on it, since I think it may be confusing tho have a different behaviour in select vs column element (that a select can be turned into using `scalar_select()`)

**Reply by Alc-Alc** [reply]
> There's a typo in the slice example.

~~@CaselIT, if you mean limit being 2 and offset being 1, then that is not a typo, I confirmed this behavior. Or did you mean something else?~~

edit: I think you meant the `.id` appearing outside the parenthesis

**Reply by CaselIT** [reply]
fixed in 6d0379f0565db1b6bf3aa7bead44d759407abadc
